crates/database/src/lib.rs

```diff
@@ -1,7 +1,25 @@
 pub mod byte_array;
+pub mod events;
 
 use byte_array::ByteArray;
-use sqlx::PgExecutor;
+use sqlx::{Executor, PgPool};
+
+// Design:
+//
+// Functions that execute multiple transactions should take `&mut PgTransaction` to indicate this
+// and to ensure that the whole function succeeds or fails together.
+// Functions that execute a single transaction should take `&mut PgConnection`.
+// We usually call the parameter `ex` for `Executor` which is the trait whose methods we use to run
+// queries.
+// This scheme allows callers to decide whether they want to use the function as part of a bigger
+// transaction or standalone. Note that PgTransaction implements Deref to PgConnection.
+// Callers do need to take care of calling `commit` on the transaction.
+//
+// For tests a useful pattern is to start a transaction at the beginning of the test, use it for all
+// queries and never commit it. This automatically allows postgres tests to run in parallel makes
+// clearing all tables at the beginning of a test obsolete.
```

Comment:
I think an important detail here is that `rollback()` gets called when a `Transaction` gets dropped and `commit()` or `rollback()` have not been called yet.
Also do we need to take precautions against dead locks when we run the tests in parallel with this pattern?

Reply:
I don't see why we would get deadlocks. As long as you the whole test runs in one transaction it should be fine. The only potential problem I see is that Postgres might make one test wait for another in which case they wouldn't really run in parallel.

Reply:
Transactions acquire locks in the database. If we write each DB test using a single big transaction and then execute all tests in parallel the many transactions might acquire locks in the DB such that 2 transactions have to wait on each other. I [read](https://dba.stackexchange.com/questions/97778/how-does-postgresql-decide-which-query-to-cancel-during-deadlock-cleanup) that postgres can detect deadlocks and abort stuck transactions at random so our CI should not get stuck completely but it might lead to tests failing randomly.
I'm not sure how likely this is to happen, though.

Reply:
That would be a problem. I thought transactions would solve this but I'm not sure anymore. If we see it happen we might have to go back to running postgres tests one by one.